crates/scroll/alloy/evm/src/tx/compression.rs

```diff
@@ -0,0 +1,214 @@
+        encoder
+            .set_parameter(CParameter::WindowLog(CL_WINDOW_LIMIT))
+            .expect("Failed to set window log");
+        encoder
+            .set_parameter(CParameter::TargetCBlockSize(target_block_size))
+            .expect("Failed to set target block size");
+        encoder.include_checksum(false).expect("Failed to disable checksum");
+        encoder.include_magicbytes(false).expect("Failed to disable magic bytes");
+        encoder.include_dictid(false).expect("Failed to disable dictid");
+        encoder.include_contentsize(true).expect("Failed to include content size");
+        encoder
+    }
+
+    /// Computes the compression ratio for the provided bytes.
+    ///
+    /// This is computed as:
+    /// `(original_size * TX_L1_FEE_PRECISION_U256) / compressed_size`
+    pub fn compute_compression_ratio<T: AsRef<[u8]>>(bytes: &T) -> U256 {
+        // Instantiate the compressor
+        let mut compressor = compressor(CL_WINDOW_LIMIT);
+
+        // Set the pledged source size to the length of the bytes and write the bytes to the
+        // compressor.
+        let original_bytes_len = bytes.as_ref().len();
+        compressor
+            .set_pledged_src_size(Some(original_bytes_len as u64))
+            .expect("failed to set pledged source size");
+        compressor.write_all(bytes.as_ref()).expect("failed to write bytes to compressor");
+
+        // Finish the compression and get the result.
+        let result = compressor.finish().expect("failed to finish compression");
+
+        // compute the compression ratio
+        let original_len = U256::from(original_bytes_len).saturating_mul(TX_L1_FEE_PRECISION_U256);
+        let compressed_len = U256::from(result.len());
+        original_len.wrapping_div(compressed_len)
+    }
+}
+
+#[cfg(not(feature = "zstd_compression"))]
+mod zstd_compression {
+    use super::*;
+
+    /// Computes the compression ratio for the provided bytes. This panics if the compression
+    /// feature is not enabled. This is to support `no_std` environments where zstd is not
+    /// available.
+    pub fn compute_compression_ratio<T: AsRef<[u8]>>(_bytes: &T) -> U256 {
+        panic!("Compression feature is not enabled. Please enable the 'compression' feature to use this function.");
+    }
+}
+
+/// A generic wrapper for a type that includes a compression ratio and encoded bytes.
+#[derive(Debug, Clone)]
+pub struct WithCompression<T> {
+    value: T,
+    compression_ratio: U256,
+    encoded_bytes: Bytes,
+}
+
+/// A trait for types that can be constructed from a transaction, its sender, encoded bytes and
+/// compression ratio.
+pub trait FromTxWithCompression<Tx> {
+    /// Builds a `TxEnv` from a transaction, its sender, encoded transaction bytes, and a
+    /// compression ratio.
+    fn from_compressed_tx(
+        tx: &Tx,
+        sender: Address,
+        encoded: Bytes,
+        compression_ratio: Option<U256>,
+    ) -> Self;
+}
```

Comment:
Where is this trait used? Is this an utility for the zk team?

Reply:
It's used in the implementation below. It is the foundation (in combination with `ToCompressed`) of how we can inject the compression factor in the guest by instantiating the `WithCompression` type:
https://github.com/scroll-tech/reth/blob/89d2bf0e21bf163b136ba137f196ee94e09a5bb6/crates/scroll/alloy/evm/src/block/mod.rs#L97-L132

This follows the same pattern as `FromTxWithEncoded` and `FromRecoveredTx` that the `BlockExecutor`'s are based on.

Reply:
missed that, thanks!